### PR TITLE
Fast Boot "Shoebox": A means to serialize application state into DOM for consumption by the browser

### PR DESCRIPTION
**Pull request by habdelra**
This PR is to serialize and deserialize application state, so that application state can be passed between the fastboot rendered app and the browser rendered app.

Todo:
- [x] Review and merge PR fastboot server serialization PR https://github.com/ember-fastboot/fastboot/pull/59
- [x] Update `fastboot-express-middleware`'s `fastboot` dependency to incorporate the PR above ^^ and bump the `fastboot-express-middleware` version
- [x] Merge @tomdale 's express middleware PR https://github.com/ember-fastboot/ember-cli-fastboot/pull/190 and rebase this PR
- [x] Update this project's `package.json` to use the bumped `fastboot-express-middleware` above ^^

In my local environment with all the `npm link`'s setup amongst all these dependent libs, all the tests are green. So fingers crossed that after all the dependencies above are completed the tests will be green.

Do note, that in order to test retrieving items from the "shoebox", we need to actually run browser tests (not to be confused with the existing mocha server tests). To that end, I had to add a bunch of ember addon boilerplate (created from `ember addon`) in order to establish the normal `ember test` qunit browser tests--so that's why all the new files...

### CODE REVIEW

**Comment by habdelra** [comment]
@tomdale ok I'm ready for your 👀 now

File: tests/dummy/app/index.html [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+  </head>
+  <body>
+    {{content-for "body"}}
+
+    <script src="assets/vendor.js"></script>
+    <script src="assets/dummy.js"></script>
+
+    {{content-for "body-footer"}}
+
+    <script type="fastboot/shoebox" data-key="key1">
```
Comment (danmcclain) [inline tests/dummy/app/index.html]: Using `id` instead of `data` attributes will likely be faster

File: app/services/fastboot.js [hunk 2]
```diff
@@ -53,5 +53,22 @@ export default Ember.Service.extend({
   deferRendering(promise) {
     Ember.assert('deferRendering requires a promise or thennable object', typeof promise.then === 'function');
     this._fastbootInfo.deferRendering(promise);
+  },
+
+  putInShoebox(key, value) {
```
Comment (danmcclain) [inline app/services/fastboot.js]: What about `shoebox.{put/get}`

File: app/services/fastboot.js [hunk 3]
```diff
@@ -53,5 +53,22 @@ export default Ember.Service.extend({
...
+    Ember.assert('putInShoebox is only invoked from the fastboot rendered application', this.get('isFastBoot'));
+    if (!this._fastbootInfo.shoebox) { this._fastbootInfo.shoebox = {}; }
+
+    this._fastbootInfo.shoebox[key] = value;
+  },
+
+  retrieveFromShoebox(key) {
+    Ember.assert('retrieveFromShoebox is only invoked from the browser rendered application', !this.get('isFastBoot'));
+    let $el = Ember.$(`script[type="fastboot/shoebox"][data-key="${key}"]`);
```
Comment (danmcclain) [inline app/services/fastboot.js]: Retrieval by id will be better optimized

File: tests/dummy/app/index.html (hunk 1 above)
Comment (habdelra) [inline tests/dummy/app/index.html]: Good idea.

File: app/services/fastboot.js (hunk 2 above)
Comment (habdelra) [inline app/services/fastboot.js]: I like it

File: app/services/fastboot.js (hunk 3 above)
Comment (habdelra) [inline app/services/fastboot.js]: Agreed

**Comment by habdelra** [comment]
@danmcclain I've updated with your suggestions. thanks!

File: app/services/fastboot.js [hunk 4]
```diff
@@ -18,18 +18,51 @@ const RequestObject = Ember.Object.extend({
...
   })
 });
 
+const Shoebox = Ember.Object.extend({
+  // careful calling `this.get`, we're overriding it, use `Ember.get` instead
+  put(key, value) {
+    Ember.assert('shoebox.put is only invoked from the fastboot rendered application', Ember.get(this, 'fastboot.isFastBoot'));
+    Ember.assert('the provided key is a string', typeof key === 'string');
+
+    let fastbootInfo = Ember.get(this, 'fastboot')._fastbootInfo;
```
Comment (danmcclain) [inline app/services/fastboot.js]: We have `get` available

File: app/services/fastboot.js [hunk 5]
```diff
@@ -18,18 +18,51 @@ const RequestObject = Ember.Object.extend({
...
+    if (!fastbootInfo.shoebox) { fastbootInfo.shoebox = {}; }
+
+    fastbootInfo.shoebox[key] = value;
+  },
+  get(key) {
+    if (Ember.get(this, 'fastboot.isFastBoot')) {
```
Comment (danmcclain) [inline app/services/fastboot.js]: Was thinking the same thing 👍

**Comment by tomdale** [comment]
This needs docs (at least API and probably guides on the FastBoot site) before we can merge.

**Pull request merged**